distributedlog-io/dlfs/src/main/java/org/apache/distributedlog/fs/DLFileSystem.java

```diff
@@ -0,0 +1,330 @@
+        try {
+            String logName = getStreamName(path);
+            if (recursive) {
+                Iterator<String> logs = namespace.getLogs(logName);
+                while (logs.hasNext()) {
+                    String child = logs.next();
+                    Path childPath = new Path(path, child);
+                    delete(childPath, recursive);
+                }
+            }
+            namespace.deleteLog(logName);
+            return true;
+        } catch (LogNotFoundException e) {
+            return true;
+        }
+    }
+
+    //
+    // Not Supported
+    //
+
+    @Override
+    public FileStatus[] listStatus(Path path) throws FileNotFoundException, IOException {
+        String logName = getStreamName(path);
+        try {
+            Iterator<String> logs = namespace.getLogs(logName);
+            List<FileStatus> statusList = Lists.newArrayList();
+            while (logs.hasNext()) {
+                String child = logs.next();
+                Path childPath = new Path(path, child);
+                statusList.add(getFileStatus(childPath));
+            }
+            Collections.sort(statusList, Comparator.comparing(FileStatus::getPath));
+            return statusList.toArray(new FileStatus[statusList.size()]);
+        } catch (LogNotFoundException e) {
+            throw new FileNotFoundException(path.toString());
+        }
+    }
+
+
+    @Override
+    public boolean mkdirs(Path path, FsPermission fsPermission) throws IOException {
+        String streamName = getStreamName(path);
+
+        // Create a dummy stream to make the path exists.
+        namespace.createLog(streamName);
+        return true;
+    }
+
+    @Override
+    public FileStatus getFileStatus(Path path) throws IOException {
+        String logName = getStreamName(path);
+        boolean exists = namespace.logExists(logName);
+        if (!exists) {
+            throw new FileNotFoundException(path.toString());
+        }
+
+        long endPos;
+        try {
+            DistributedLogManager dlm = namespace.openLog(logName);
+            endPos = dlm.getLastTxId();
+        } catch (LogNotFoundException e) {
+            throw new FileNotFoundException(path.toString());
+        } catch (LogEmptyException e) {
+            endPos = 0L;
+        }
+
+        // we need to store more metadata information on logs for supporting filesystem-like use cases
+        return new FileStatus(
+            endPos,
+            false,
+            3,
+            dlConf.getMaxLogSegmentBytes(),
+            0L,
+            makeAbsolute(path));
+    }
+
+
+    @Override
+    public boolean rename(Path src, Path dst) throws IOException {
+        String srcLog = getStreamName(src);
+        String dstLog = getStreamName(dst);
+        namespace.renameLog(srcLog, dstLog);
+        return true;
+    }
+
+    @Override
+    public boolean truncate(Path f, long newLength) throws IOException {
```

Comment:
How can this even be supported without reading the whole log and rewriting? Is this used much in any case? Seems to me like a map reduce specific thing.

Reply:
truncate(f, newLength) is a posix filesystem semantic. I see it is used in multiple places -- when people want to delete potential-corrupted data from file (e.g. due to application crashes), they use truncate.

in dlog, we support truncate the head. in filesystem, truncate is truncating the tail. it is actually easy to support - whenever a truncate is issued, it should close active segment (ledger), based on the length to find what segments can be fully deleted and what segments can be partially deleted. for segments that can be fully deleted, just delete the ledgers and for segments that can be partially deleted, update the end position (dlsn) in the segments' metadata.

all the metadata status and mechanism are sorted of ready in dlog, since it supports truncate the head. adding truncate to tail can use those statuses.

Reply:
Create an issue for this - https://github.com/apache/distributedlog/issues/232